### PR TITLE
Add tone generation for ESP32

### PR DESCRIPTION
**Pull request by EvilGremlin**
### Description
Add tone generation for ESP32.
Bump up platform version, becuse why not? It looks like nothing's broken.
Someone who actually have an idea how it all works, please check.

### Benefits
#define SPEAKER now works
Yay M300 we have cool tunes!

### CODE REVIEW

File: platformio.ini [hunk 1]
```diff
@@ -1482,12 +1482,13 @@ build_flags       = ${stm32_flash_drive.build_flags}
...
 [env:esp32]
-platform      = espressif32@1.11.2
+platform      = espressif32@2.1.0
 board         = esp32dev
 build_flags   = ${common.build_flags} -DCORE_DEBUG_LEVEL=0
 src_filter    = ${common.default_src_filter} +<src/HAL/ESP32>
 lib_ignore    = NativeEthernet
-upload_speed  = 115200
+upload_speed  = 500000
+monitor_speed = 250000
```
Comment (thinkyhead) [inline platformio.ini:1491]: Any ESP configuration examples that have the lower baud rate should be updated to this higher one. Or, this should be left at 115200 if the configurations are better to retain the "more reliable" lower baud rate.

Comment (vivian-ng) [reply platformio.ini:1491]: I recommend to leave it at 115200 for more stable and reliable operation. Higher baud rates may or may not work on some boards depending on which chip they are using for the USB interface.

Comment (thinkyhead) [reply platformio.ini:1491]: Hmm, looks like no examples are set up, so… off we go.

Comment (EvilGremlin) [reply platformio.ini:1491]: Well, pio.ini can be excluded then, that's why i did separate commit.

**Comment by thinkyhead** [comment]
> Still kinda wurbly…

If the amplitude is irregular, then possibly there's less current to the speaker during certain activities, but most boards should have that pretty well isolated. But if the frequency is uneven, we can try boosting the timer priority (or choosing a timer index with a naturally higher priority) to see if that helps. I'm not sure exactly where to make that change… yet.

Note that currently our ESP32 implementation of `HAL_timer_disable_interrupt` doesn't do anything. I assume that is by design, and I doubt that has any negative effect on timing.

**Comment by EvilGremlin** [comment]
Ah, TOGGLE is nice, missed that.
There is noticeable stutter every half second or so. Gotta try without webui... But i need to improve circuit first as simple NPN to ground isn't proper driver for piezo beeper.

**Comment by simon-jouet** [comment]
So I just saw this MR. Using timers in the ESP32 works okay but the number of interrupts can cause issues with the WiFi (or vice versa depending in the priorities). I think the tone generation should have been done with the ledc code (like PWM) to use the hardware to generate the tone. I think using the api `ledc_set_fade_with_time` most of the tone generation can be done in hardware.

**Comment by EvilGremlin** [comment]
Yep, it's just a quick and dirty copypaste, i never expected it to be good. But i made it like that so i could use i2s, and it kinda works. I will use 555 to make pulse lenghts actually usable for sound and report back.

**Comment by vivian-ng** [comment]
Sorry I haven't been trying out the latest changes to Marlin bugfix. Didn't realize that the tone generation uses I2S. Will this affect the I2S stepper stream? If it affects I2S stepper stream, the ledc idea may be a more compatible solution that doesn't conflict with I2S or Wifi.

**Comment by EvilGremlin** [comment]
Who knows? I need to get it printing first, to see if anything affected. Then again, it won't really play tunes while actually printing and menu beeps may be not sufficient for testing.

**Comment by simon-jouet** [comment]
So I don't think using I2S for the tone generation is the right thing to do (it will require a software pulse generator for this to happen). Just using the ledc driver and a normal GPIO is the best approach I think. I might have a look at some point but I've never used a buzzer on any of my controllers so this hasn't been very high prio

**Pull request merged**